File: core/src/main/java/pcb/uwu/amount/derived/optics/Illuminance.java
package pcb.uwu.amount.derived.optics;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.optics.IlluminanceUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class Illuminance extends CompositeUnitAmount<IlluminanceUnit> {

	// region constructors

	public Illuminance(Number value, IlluminanceUnit unit) {
		super(value, unit);
	}

	public Illuminance(Number value, Magnitude magnitude, IlluminanceUnit unit) {
		super(value, magnitude, unit);
	}

	public Illuminance(String value, IlluminanceUnit unit) {
		super(value, unit);
	}

	public Illuminance(String value, Magnitude magnitude, IlluminanceUnit unit) {
		super(value, magnitude, unit);
	}

	public Illuminance(BigDecimal value, IlluminanceUnit unit) {
		super(value, unit);
	}

	public Illuminance(BigDecimal value, Magnitude magnitude, IlluminanceUnit unit) {
		super(value, magnitude, unit);
	}

	public Illuminance(BigDecimalAmount amount, IlluminanceUnit unit) {
		super(amount, unit);
	}

	public Illuminance(BigDecimalAmount amount, Magnitude magnitude, IlluminanceUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Illuminance plus(UnitAmount<IlluminanceUnit> other, MathContext mathContext) {
		return new Illuminance(plusAmount(this, other, mathContext), getUnit());
	}

	@Override
	public Illuminance minus(UnitAmount<IlluminanceUnit> other, MathContext mathContext) {
		return new Illuminance(minusAmount(this, other, mathContext), getUnit());
	}

	@Override
	public Illuminance multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Illuminance(multipliedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public Illuminance dividedBy(BigDecimal other, MathContext mathContext) {
		return new Illuminance(dividedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public Illuminance in(IlluminanceUnit unit) {
		return new Illuminance(getAmountIn(this, unit), unit);
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/amount/derived/optics/Illuminance.kt
package pcb.uwu.amount.derived.optics

import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.optics.IlluminanceUnit
import pcb.uwu.utils.UnitAmountUtils

open class Illuminance : CompositeUnitAmount<IlluminanceUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: IlluminanceUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: IlluminanceUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    override fun plus(amount: UnitAmount<IlluminanceUnit>) =
        Illuminance(amount = this.amount + (amount to this.unit).amount,
                    unit = this.unit)

    override fun minus(amount: UnitAmount<IlluminanceUnit>) =
        Illuminance(amount = this.amount - (amount to this.unit).amount,
                    unit = this.unit)

    override fun times(number: Number) =
        Illuminance(amount = this.amount * number,
                    unit = this.unit)

    override fun div(number: Number) =
        Illuminance(amount = this.amount / number,
                    unit = this.unit)

    override fun to(unit: IlluminanceUnit) =
        Illuminance(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                    unit = unit)

    // endregion
}